src/runtime/importhook.cs

```diff
@@ -141,189 +128,177 @@ internal static void SaveRuntimeData(RuntimeDataStorage storage)
+
+            var loader = Runtime.PyDict_GetItemString(mod_dict, "DotNetLoader").DangerousGetAddressOrNull();
+            Runtime.XIncref(loader);
+
+            // Add the classes to the module
+            // PyModule_AddObject steals a reference only on success
+            if (Runtime.PyModule_AddObject(import_hook_module, "DotNetLoader", loader) != 0)
             {
-                Runtime.PyDict_Update(py_mod_dict, clr_dict);
+                Runtime.XDecref(loader);
+                throw new PythonException();
             }
 
-            // find any items from the from list and get them from the root if they're not
-            // already in the module dictionary
-            if (fromList != null)
+            var finder = Runtime.PyDict_GetItemString(mod_dict, "DotNetFinder").DangerousGetAddressOrNull();
+            Runtime.XIncref(finder);
+            if (Runtime.PyModule_AddObject(import_hook_module, "DotNetFinder", finder) != 0)
             {
-                if (Runtime.PyTuple_Check(fromList))
-                {
-                    using var mod_dict = new PyDict(py_mod_dict);
-                    using var from = new PyTuple(fromList);
-                    foreach (PyObject item in from)
-                    {
-                        if (mod_dict.HasKey(item))
-                        {
-                            continue;
-                        }
-
-                        var s = item.AsManagedObject(typeof(string)) as string;
-                        if (s == null)
-                        {
-                            continue;
-                        }
-
-                        ManagedType attr = root.GetAttribute(s, true);
-                        if (attr == null)
-                        {
-                            continue;
-                        }
+                Runtime.XDecref(finder);
+                throw new PythonException();
+            }
 
-                        Runtime.XIncref(attr.pyHandle);
-                        using (var obj = new PyObject(attr.pyHandle))
-                        {
-                            mod_dict.SetItem(s, obj);
-                        }
-                    }
-                }
+            // Set as a sub-module of clr.
+            Runtime.XIncref(import_hook_module);
+            if(Runtime.PyModule_AddObject(py_clr_module, "loader", import_hook_module) != 0)
+            {
+                Runtime.XDecref(import_hook_module);
+                throw new PythonException();
             }
-            Runtime.XIncref(py_clr_module);
-            return NewReference.DangerousFromPointer(py_clr_module);
+
+            // Finally, add the hook to the meta path
+            var finder_inst = Runtime.PyDict_GetItemString(mod_dict, "finder_inst").DangerousGetAddressOrNull();
+            Runtime.XIncref(finder);
+            var metapath = Runtime.PySys_GetObject("meta_path");
+            Runtime.PyList_Append(metapath, finder);
```

Comment:
These four lines look fishy: you don't use finder_inst and you are doing an incref on finder for no obvious reason.

Reply:
oh.. that should be `finder_inst`, not finder

Reply:
BTW, it makes sense to add an overload for `PyList_Append` with proper reference types, then avoid manual refcounting.

Same for `PySys_GetObject` - you can rename the original one to `Dangerous_PySys_GetObject` to avoid refactoring other call sites.

This is a NIT though. Should come up with a better plan to switch everything.